### DISCUSSION TITLE
"Control Surface Studio" discussion

### DISCUSSION THREAD

**Discussion by markusschloesser** [markusschloesser/MackieC4_P3#96] [Ideas]
[BilldarBagdar](https://github.com/BilldarBagdar)
[2 days ago](https://github.com/markusschloesser/MackieC4_P3/discussions/95#discussioncomment-3854851)
Collaborator
Hi Markus,

I hope rather than starting a new topic, that commenting here is a good enough place to mention this idea.

I bought a development program yesterday called [Control Surface Studio](https://remotify.io/product/control-surface-studio) (desktop version) based on some of the associated YouTube videos, but mostly because I have a crazy idea about programming a (now freeware) application called [Musicwonk ](http://algoart.com/)to be a kind of "Algorithmic DJ" as well as "Algorithmic keyboard player". I.E. I want to write a Live "remote script" so Live will treat Musicwonk like a control surface (and of course also make Musicwonk patches that behave like an "Algorithmic DJ" operating itself as both a control surface and a "keyboard player").

I think CSS will work great helping me realize the above idea. But I don't think I really know enough about how everything works together in CSS yet to just take off marching to my own drummer on some crazy vector like that. I'm thinking about first attempting to convert this C4 script into the CSS "format" and continuing to work on it in CSS. This is the idea I'm mentioning.

Have you heard of Control Surface Studio? I suspect the v2/v3 framework question posed in this thread would become moot, more or less.

Would you be opposed to such a (radical?) change in direction for our MackieC4 script collaboration project? Actually, I don't think opposed is the right word because I'm not sure anything would necessarily change here at GitHub. I don't think CSS produces "source code" artifacts like would normally get archived in a git repo.

One of the things I like about CSS is the "live object model" is menu driven for script writers, the events are all listed in nested menus so you just pick the one you want.

**Comment by markusschloesser** [comment]
[BilldarBagdar](https://github.com/BilldarBagdar)
[yesterday](https://github.com/markusschloesser/MackieC4_P3/discussions/95#discussioncomment-3863589)
Collaborator
I uploaded version 0.0.1 of a Mackie C4 "Control Surface Template" file for use with CSS to their forum. The way I understand the situation now, it means you or anyone can freely download that template for use with CSS desktop, or sign up for the free account at remotify.io and use that template with either the paid or complimentary online version of CSS. Either way, you or anyone can keep and use any "scripts" you create using whatever version.

Aside from the super tight "script development" feedback loop evident in all the videos, I think the CSS product feature that really sealed the deal for me was showcased in the video on "[reactions](https://www.youtube.com/watch?v=iYtotPDoa3k)". In other words, the topic of the video is "Listeners run code in reaction to being notified when an Event has occurred". All you need to think about while developing with CSS is how you want your device and Live to interact. You don't need to think about any of the underlying/supporting python scaffolding/infrastructure unless until you hit a wall, and then "custom python code" options remain available.

I'm not sure I (we?) will be able to completely replicate all behavior from this script in a CSS C4 script immediately, I mean without additional CSS features being added by the developer over time. But I am sure most of the functionality can be replicated now, including for example, updating the LCD screens via SYSEX in response to events. I'm not yet sure if CSS will natively allow, for example, the "master volume" metering this script can display. I don't know for sure if I can make a CSS script run code in response to an "on display update" Event, but I think I can. I also think I can simplify what happens when the "on display update" method runs (i.e. when that Listener reacts to that event)

Proof, however, will be in the pudding. We'll see how far I actually get.

**Reply by markusschloesser** [reply]
Hey @BilldarBagdar sounds quite fascinating! I heard about CSS a couple of times already, but the last time I looked at it, it didn't do what I wanted, so I stopped looking. I definitely wish good luck with it! :-)
And I am definitely interested in the outcome of this! Can you share a script out of Css?

I will keep working on the current version of the MackieC4 script, improve it, add functionality etc. And then I wanna migrate to the v3 framework at some point.

**Reply by BilldarBagdar** [reply]
Hey Thanks for moving the discussion.  I will keep working on this script too, ha, as much as ever.  I hit a CSS showstopper almost immediately and no response yet on their forum.  So, IDK, maybe the developer is on vacation this week.

The one script I've been able to "install" successfully and review appears to be normal, advanced looking Python code, but that was just a quick test after my first C4 script wouldn't install.  So, I'm pretty sure I would be able to share whatever scripts I develop.  Sharing Templates and Scripts is encouraged on their website.  

All I really did was map the Track L, Track R, Slot Up, and Slot Down buttons to control "Highlight" and "Session Box" Navigation in "Channel Strip" mode and the "Channel Strip" button with LED feedback.  But the script wouldn't "compile" when I tried to test just that much, so now I'm waiting for the developer to respond, hopefully with a fix.  I was able to use the software long enough though to begin enjoying the workflow and seeing the powerful possibilities.  It's point and click programming for the most part.  The C4 encoder rings are going to need "custom code" treatment (raw midi messages) to send the correct Single Dot, Boost Cut, Spread, etc. settings.  But I'm not that far yet, and then there's the LCD screen data...  This script works pretty well though, no rush to move away, and I still like learning Python.

I also learned enough playing around learning to write a C4 script to begin to see the kind of Live Control scripting "Algorithmic DJ" I can make out of Musicwonk patches.  That appears to be an exciting area for further exploration, to me anyway.
